app/models/user.rb

```diff
@@ -206,7 +207,42 @@ def self.authenticator(username = nil)
   end
 
   def self.authenticate(username, password, request = nil, options = {})
-    authenticator(username).authenticate(username, password, request, options)
+    user = authenticator(username).authenticate(username, password, request, options)
+    user.try(:link_user_to_session, request)
+
+    user
+  end
+
+  def link_user_to_session(request)
+    return unless request
+
+    session_id = request.session_options[:id]
+    sessions << Session.find_or_create_by(:session_id => session_id)
+  end
+
+  def broadcast_revoke_sessions
+    if Settings.server.session_store == "cache"
+      MiqQueue.broadcast(
+        :class_name  => self.class.name,
+        :instance_id => id,
+        :method_name => :revoke_sessions
+      )
+    else
+      # If using SQL or Memory, the sessions don't need to (or can't) be
+      # revoked via a broadcast since the session/token stores are not server
+      # specific, so execute it inline.
+      revoke_sessions
+    end
+  end
+
+  def revoke_sessions
+    current_sessions = Session.where(:user_id => id)
+    ManageIQ::Session.revoke(current_sessions.map(&:session_id))
+    current_sessions.destroy_all
```

Comment:
Wondering if this should be in a transaction?

Reply:
Probably not a bad idea.

Reply:
Actually, I am going to hold off on this.

While a transaction would be nice, unfortunately, there is a possibility it will be not function as expected, since `ManageIQ::Session` could have a `.store` of `memcached`, so the transaction wouldn't be "fully reverted" if we were to wrap both `.revoke` and `.destroy_all` in a single transaction.

I think the `.destroy_all` would be in it's own transaction, but I could be wrong so I don't know if there is any value of wrapping just that line.